keras_nlp/models/xlnet/xlnet_encoder.py

```diff
@@ -0,0 +1,349 @@
+        )
+
+        self.segment_attention_bias = self.add_weight(
+            shape=(self.num_heads, self.head_dim),
+            initializer=self.bias_initializer,
+            trainable=True,
+            name="segment_attention_bias",
+        )
+
+        self.segment_encoding = self.add_weight(
+            shape=(2, self.num_heads, self.head_dim),
+            initializer=self.kernel_initializer,
+            trainable=True,
+            name="segment_encoding",
+        )
+
+        super().build(input_shape)
+
+    def call(
+        self,
+        output_h,
+        attn_mask_h,
+        attn_mask_g,
+        pos_emb,
+        seg_mat,
+        output_g=None,
+        mems=None,
+        target_mapping=None,
+    ):
+        # rel_attn
+        attn_out_h, attn_out_g = self.relative_attention(
+            content_stream=output_h,
+            query_stream=output_g,
+            content_attention_mask=attn_mask_h,
+            query_attention_mask=attn_mask_g,
+            relative_position_encoding=pos_emb,
+            content_attention_bias=self.content_attention_bias,
+            positional_attention_bias=self.positional_attention_bias,
+            segment_attention_bias=self.segment_attention_bias,
+            segment_matrix=seg_mat,
+            segment_encoding=self.segment_encoding,
+            target_mapping=target_mapping,
+            state=mems,
+        )
+
+        attn_out_h = self.dropout_attn(attn_out_h)
+        attn_out_h = attn_out_h + output_h
+        attn_out_h = self.layer_norm(attn_out_h)
+
+        if attn_out_g is not None:
+            attn_out_g = self.dropout_attn(attn_out_g)
+            attn_out_g = attn_out_g + output_g
+            attn_out_g = self.layer_norm(attn_out_g)
+
+        # feed-forward
+        ff_out_h = attn_out_h
+        ff_out_h = self.feedforward_intermediate_dense(ff_out_h)
+        ff_out_h = self.activation_function_ff(ff_out_h)
+        ff_out_h = self.dropout_ff(ff_out_h)
+        ff_out_h = self.feedforward_output_dense(ff_out_h)
+        ff_out_h = self.dropout_ff(ff_out_h)
+        ff_out_h = self.layer_norm_ff(ff_out_h + attn_out_h)
+
+        if attn_out_g is not None:
+            ff_out_g = attn_out_g
+            ff_out_g = self.feedforward_intermediate_dense(ff_out_g)
+            ff_out_g = self.activation_function_ff(ff_out_g)
+            ff_out_g = self.dropout_ff(ff_out_g)
+            ff_out_g = self.feedforward_output_dense(ff_out_g)
+            ff_out_g = self.dropout_ff(ff_out_g)
+            ff_out_g = self.layer_norm_ff(ff_out_g + attn_out_g)
+
+            return ff_out_h, ff_out_g
+
+        return ff_out_h, None
+
+    def compute_output_shape(
+        self,
+        output_h_shape,
+        pos_emb_shape,
+        attn_mask_h_shape,
+        attn_mask_g_shape,
+        seg_mat_shape,
+        output_g_shape=None,
+    ):
+        return [output_h_shape, output_h_shape]
+
+
+class XLNetEncoderBlockPreprocessingLayer(keras.layers.Layer):
```

Comment:
I would keep this as it's own file, and not name it a Preprocessor layer (that means something different in our library). Maybe two separate layers `XLNetAttentionMaskLayer`, `XLNetSegmentMatrixLayer`?

Reply:
created two layers - `XLNetAttentionMaskLayer`, `XLNetSegmentMatrixLayer` as you instructed.
Please let me know if I should put them in a separate file (`xlnet_encoder_logic.py`) or not?

Reply:
I think what you have here is fine. As a general rule, a one-to-one filename -> classname correspondence is what we usually do. We make exception for really small helper classes on occasion.

But I think we can merge as is. None of these internal layers are exported.